### ISSUE TITLE
Serial via GPIO

### ISSUE THREAD

**Issue by maertz** [nebrius/raspi-serial#4]
I'm just wondering if it's possible to use this package to send data over a GPIO.

**Comment by nebrius** [comment]
This package does not support serial over GPIO. AFAIK it's not possible to do this on the RPi sadly.

**Comment by maertz** [comment]
Oh alright, that's a bummer but there might be a way. Actually I've built something to get this working, I was just hoping to use this package instead.

I've used pigpio with [wave_add_serial](http://abyz.me.uk/rpi/pigpio/python.html#wave_add_serial) and [bb_serial_read](bb_serial_read_open) to communicate with my [mp3 modul](https://www.dfrobot.com/product-1121.html).

**Comment by nebrius** [comment]
Oh nice, I didn't know about that capability!

We would first need to get support added in to https://github.com/fivdi/pigpio. Then, we can create a new module called raspi-soft-serial that implements that functionality, similar to how https://github.com/nebrius/raspi-soft-pwm implements software-based PWM.

@fivdi do you have any thoughts on exposing software serial?

**Comment by fivdi** [comment]
@nebrius There's an open issue related to the topic at https://github.com/fivdi/pigpio/issues/40. It's not something that I intend adding in the near future.

@maertz The Raspberry Pi has a real serial port available on [pins 8 and 10 of the GPIO header](https://www.raspberrypi-spy.co.uk/wp-content/uploads/2012/06/Raspberry-Pi-GPIO-Layout-Model-B-Plus-rotated-2700x900.png). This serial port should be supported by raspi-serial. There may be a few hoops to go through but it should work. Can this serial port not be used? Depending on the model a Raspberry Pi also has up to 4 USB ports. In combination with a USB to serial cable these USB ports could also be used.

**Comment by nebrius** [comment]
To add to what @fivdi said, take a look at the [constructor options](https://github.com/nebrius/raspi-serial#new-serialoptions) to see how to use a serial port other than the default one on pins 8 and 10.

**Comment by maertz** [comment]
Thanks for the quick support. I was able to make it work with the normal TX/RX pins, all good so far. Unfortunately I'm forced in my project to use the SPI pins (MOSI/MISO), therefore I've enabled SPI and tried to hook up `/dev/spidev0.1` but It was not able to apply the custom baudrate.

I found [this project](https://github.com/themrleon/RpiSoft-UART) to create a "virtual serial interface" as far as I understood but I was not able to compile it to test this. In theory this might work :)

**Comment by fivdi** [comment]
I hope this is not a stupid question but are you using the SPI pins for SPI or are you using them for something else?

When you say you are not able to apply the custom baudrate do you mean the baudrate of the serial port or the baudrate of the SPI bus?

**Comment by maertz** [comment]
It might be a little bit confusing. Currently I'm using the SPI pins as regular GPIO pins to bitbang UART commands to communicate with my mp3 module (it could be any GPIO pin but I'm forced by my custom board to use the SPI ones). This works great, is consistent as far as the cpu pressure is low but has a side effect on pin 18. It looks like the PCM clock changes during the bitbanging, therefore I'm looking for an alternate.

I've tried to use actual SPI with several frequencies to simulate UART to achieve the same but with no luck.

In my latter attempt I've tried to connect raspi-serial with my SPI bus which didn't work, probably for a reason but I've actually no idea what I'm doing – I'm just hacking around to make it work.
